Added mesh frustum culling

otsakir opened the pull request:
Frustum culling support for mesh component (sphere algorithm)

Fixes #6455

otsakir committed:
first working implementation of mesh frustum culling (#6455)

otsakir committed:
progress on mesh frustum culling (#6455)

* Support only float mesh but  don't crush with other types
* Versioning for buffer bound info. Do not recalculate.
* Apply culling only of bound info available
* Check for buffer/stream validity in new functionality

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -616,4 +621,70 @@ namespace dmBuffer
     }
+
+
+    Result UpdateBounds(HBuffer hbuffer, dmhash_t stream_name) {
```
We said we should calculate and set this from the mesh component.

otsakir replied:
yup we did :-). 

I tried that first but there were to many dependencies on private parts of the code in buffer.cpp and moved this to buffer.cpp to make it work.

Let me try this again and discuss not that things get clearer.

otsakir replied:
Ok, here is the obstacle i decided to move this (UpdateBounds()) to buffer.cpp instead of comp_mesh.cpp for:

`GetBuffer()` requires a `g_BufferContext` to be able to get the internal buffer from an HBuffer. This is private stuff only available in buffer.cpp.

Thoughts ?

otsakir replied:
nevermind. I moved UpdateBounds() to buffer.cpp. Used GetStrean() without problems.

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -616,4 +621,70 @@ namespace dmBuffer
     }
+
+
+    Result UpdateBounds(HBuffer hbuffer, dmhash_t stream_name) {
```
I'd rather see this in 3 parts.
1) Get the data array and length
2) Calculate the bounding min + max from the data array
3) Set the bounding min+max on the buffer

otsakir replied:
You mean better separate the function implementation into three sections, right ?

Or have three separate functions ?

otsakir replied:
Let's discuss on the updated version of the code. Many things have changed.

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -616,4 +621,70 @@ namespace dmBuffer
+        }
+        Buffer::Stream* stream = GetStream(buffer, stream_name);
+        if (stream == 0x0) {
+            return RESULT_STREAM_MISSING;
```
How is the missing stream case handled?
Will the bounding info be set so that it's still visible?

otsakir replied:
This has been moved to comp_mesh.cpp and a an error has been reported. 

Closing this as it's out of date.

JCash commented on engine/dlib/src/dlib/buffer.h:
```diff
@@ -22,6 +22,15 @@ namespace dmBuffer
 
 // Used by the engine
 
+struct Bounds3D {
```
I'd rather we stick with our vector values for now (dmVMath::Point3)

JCash commented on engine/dlib/src/dlib/buffer.h:
```diff
@@ -74,6 +83,10 @@ Result GetStreamOffset(HBuffer buffer, uint32_t index, uint32_t* offset);
 
 Result CalcStructSize(uint32_t num_streams, const StreamDeclaration* streams, uint32_t* size, uint32_t* offsets);
 
+Result UpdateBounds(HBuffer hbuffer, dmhash_t stream_name);
+
+Result GetBounds(HBuffer hbuffer, Bounds3D** bounds);
```
Replace with a SetBounds(hbuffer, min, max);

otsakir replied:
Ok for min/max as points. But how will `SetBounds()` know which stream to process ?

otsakir replied:
The interface Get/SetBounds has been reworked as suggested. I'll resolve this.

JCash commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -447,7 +450,8 @@ namespace dmGameSystem
                     info->m_Version = component.m_BufferVersion;
 
-                    CopyBufferToVertexBuffer(br->m_Buffer, info->m_VertexBuffer, br->m_Stride, br->m_ElementCount, dmGraphics::BUFFER_USAGE_DYNAMIC_DRAW);
+                    dmBuffer::UpdateBounds(br->m_Buffer, dmHashString64("position"));
```
Also handle the case where it fails to calculate a bounding box.
(should we log an error? should it still be visible?)

otsakir replied:
Errors are logged by the UpdateMeshBounds(). Resolving this as out of date.

JCash commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -729,6 +733,46 @@ namespace dmGameSystem
+            dmBuffer::Bounds3D* bounds = 0x0;
+            dmBuffer::Result r = dmBuffer::GetBounds(br->m_Buffer, &bounds);
+
+            if (r != dmBuffer::RESULT_OK) {
+                return;
+            }
```
This will break the frustum culling for the rest of the items?

otsakir replied:
yes, sloppy stuff. It has been fixed.

otsakir committed:
Reworked mesh culling implementation after draft PR review (#6455)

otsakir committed:
Use stream name from resource for position vertices if available (#6455)

otsakir committed:
Error handling and code cleanup (#6455)

otsakir committed:
Better error handling (#6455)

min/maxBounds initialized to arbitrary value to help detect if they are updated

otsakir commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -303,6 +305,59 @@ namespace dmGameSystem
+        }
+
+        // set up initial values
+        dmVMath::Point3 boundsMin = BOUNDS_MIN_EMPTY;
```
That's the best solution i could come down with considering all other requirements for this feature.

boundsMin/Max are initialized to a _specific_ far away value. In case of error, the bounds are left to their initial values.  Later, in RenderListFrustumCulling(), if the boundsMin/Max have not been updated entry's m_Visibility is not udated.

otsakir committed:
Minor fixes (6455)

JCash commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -729,6 +791,53 @@ namespace dmGameSystem
+            }
+
+            // are there any actual bounds set ? (checking if boundxMin still contains its initial values should be enough)
+            if (boundsMin.getX() == BOUNDS_MIN_EMPTY.getX() && boundsMin.getY() == BOUNDS_MIN_EMPTY.getY() && boundsMin.getZ() == BOUNDS_MIN_EMPTY.getZ()) {
+                continue;
+            }
```
I wonder if this may be more "generic":

```c++
bool IsAABBValid(P3 min, P3 max) {
    return !(min.getX() > max.getX() ||
        min.getY() > max.getY() ||
        min.getZ() > max.getZ());
}
```

I like the fact that the calculation is "contained" and uses it's own local variables, rather than loading the global constant from memory.

Also, I think it's easier to communicate as an "API" to our users: "If the bounding box is invalid, the object will not be culled."

otsakir replied:
Both arguments stand. I like the suggestion too.

My only concert is whether we can assert that buffer bounds have been initialized if we choose this path. It can easily happen to have an invalid uninitialized bounding box with bound_max > bound_min. We'll take this a a valid one which is wrong. 

In the existing implementation it is highly improbable to have bounds set to +-1e10 out of bad luck. So, at least we won't worry about this.

otsakir replied:
i would place this in dlib/buffer.h. Make it private for now.

otsakir replied:
It's not really a buffer thing but since the bounds are initialized in buffer.cpp, it seems ok to me.

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -56,6 +57,10 @@ namespace dmBuffer
         uint8_t  m_NumStreams;
+
+        // two point min/max definition of a AABB
+        dmVMath::Point3 boundsMin;
+        dmVMath::Point3 boundsMax;
```
We should initialize them too, when the buffer is created.
I don't think they need certain constants, just initialize it to an "invalid" box.

otsakir replied:
This seems to cancel my previous comment. With proper initialization we don't have something to worry about.

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -616,4 +621,29 @@ namespace dmBuffer
         buffer->m_ContentVersion++;
         return RESULT_OK;
     }
+
+    Result SetBounds(HBuffer hbuffer, const dmVMath::Point3& min, const dmVMath::Point3& max) {
+        Buffer* buffer = GetBuffer(g_BufferContext, hbuffer);
+        if (!buffer) {
+            return RESULT_BUFFER_INVALID;
+        }
+
+        buffer->boundsMin = min;
+        buffer->boundsMax = max;
+
+
+        return RESULT_OK;
+    }
+
+    Result GetBounds(HBuffer hbuffer, dmVMath::Point3& min, dmVMath::Point3& max) {
+        Buffer* buffer = GetBuffer(g_BufferContext, hbuffer);
+        if (!buffer) {
+            return RESULT_BUFFER_INVALID;
+        }
+
+        min = buffer->boundsMin;
+        max = buffer->boundsMax;
+
+        return RESULT_OK;
+    }
```
👍

JCash commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -729,6 +791,53 @@ namespace dmGameSystem
+            }
+
+            // get center of bounding box in local coords
+            float center_x = (boundsMax.getX() + boundsMin.getX())/2;
+            float center_y = (boundsMax.getY() + boundsMin.getY())/2;
+            float center_z = (boundsMax.getZ() + boundsMin.getZ())/2;
```
Doing this calculation each frame, I wonder if the data we stored is wrong?
Should we store "center+half_extent" instead?

otsakir replied:
ok then. I'll keep that in mind.

otsakir commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -340,6 +395,10 @@ namespace dmGameSystem
             CreateVertexBuffer(world, br, component->m_BufferVersion);
+
+            // process position stream in buffer to find bounding box
+            uint64_t stream_name = (component->m_Resource->m_PositionStreamId != 0) ? component->m_Resource->m_PositionStreamId : dmHashString64("position"); // use stream name (hash) from resource and fallback to "position" if this is not available
```
Right, so set the "default" name at resource level instead of component level.

otsakir commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -303,6 +305,59 @@ namespace dmGameSystem
+            float y = ((float*)position_p)[1];
+            float z = ((float*)position_p)[2];
+
+            boundsMax.setX( dmMath::Max(x, boundsMax.getX()) );
+            boundsMin.setX( dmMath::Min(x, boundsMin.getX()) );
+            boundsMax.setY( dmMath::Max(y, boundsMax.getY()) );
+            boundsMin.setY( dmMath::Min(y, boundsMin.getY()) );
+            boundsMax.setZ( dmMath::Max(z, boundsMax.getZ()) );
+            boundsMin.setZ( dmMath::Min(z, boundsMin.getZ()) );
```
Ok. I'll directly use Vectormath::Aos::maxPerElem() then.

otsakir commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -446,6 +505,9 @@ namespace dmGameSystem
                     info->m_Version = component.m_BufferVersion;
+                    // process position stream in buffer to find bounding box
+                    uint64_t stream_name = (component.m_Resource->m_PositionStreamId != 0) ? component.m_Resource->m_PositionStreamId : dmHashString64("position"); // use stream name (hash) from resource and fallback to "position" if this is not available
+                    UpdateMeshBounds(br->m_Buffer, stream_name);
```
Yes, they will be. 

But dmBuffer()::SetBounds() is supposed to be a private method right ? User cannot call it. For now. 

Later, when we will make the api public we can:

1. Settle `dmBuffer::SetBounds(min,max)` api to `SetBounds(min,max, force=false)`. That is, add one more parameter for flexibility.
2. Add a version field (say buffer.m_BoundsVersion) in the buffer meta data. It will keep buffer version at the time SetBounds() was called. SetBounds() will check if the bound info it has is up to date using the version identifier and will not update the bounds if it is.
2. The component logic will stay the same. Call SetBounds() as it does currently. `force` will be false.
3. When the user wants to explicitly set the bounds, he/she calls SetBounds() with `force` set to true. 

@JCash

JCash replied:
I think the #2 is the only viable option.
The "force" flag doesn't really work I'd say.

I recommend you do a test in C++.

JCash replied:
I didn't have a good example handy, but this looks like it might work:
https://forum.defold.com/uploads/default/original/3X/f/b/fbb9c4abd3ed239f6066f912147c50744bdafa90.zip

Very simple test, creating a quad.
Test by setting bounds on the buffer directly, e.g. exaggerate the bounds, and then check that it's intact when arriving at the culling step.

otsakir replied:
Thanks.

In line with our discussion i've updated the PR with the following:

1. New m_BoundingVersion field in `Buffer`. It gets m_ContentVersion's value when SetBounds() is called.
2. Made Get/SetBounds() public i.e. moved them to dmsdk/dlib/buffer.h
3. Introduced (private) dmBuffer::BoundsUpToDate(). It tells if the min/max bounds in Buffer are up-to-date by comparing the two version fields. I'll explain why we need this in a while.
4. comp_mesh.cpp::UpdateSetBounds() will only calculate and update buffer bounds if they are out of date.

#### Why BoundsUpToDate()

Even with the newly added Buffer.m_BoundingVersion it's not possible to avoid the automatic re-calculation of bounds triggered by the component. Suppose the user updates the buffer from C++, calls dmBuffer::UpdateContentVersion() and dmBuffer::SetBounds() accordingly. The mesh component won't know. It will detect the content version change in the buffer and (a) calculate new bounding box on its own, (b) try to update the buffer with this bounding box. We could avoid (b) if SetBounds() itself checks if there is a version mismatch inside the buffer. We can't avoid (a) though. 

That's the reason i added BoundsUpToDate(). Give some insight to the component so it won't mess with the bounds explicitly given by the user. I put it into dlib/buffer.h. That's not pubic API right ?

otsakir commented on engine/gamesys/src/gamesys/components/comp_mesh.cpp:
```diff
@@ -303,6 +305,59 @@ namespace dmGameSystem
+        dmBuffer::ValueType stream_type;
+        uint32_t stream_type_count;
+        GetStreamType(hbuffer, stream_name, &stream_type, &stream_type_count); // hbuffer and stream_name have been validated
+        if (stream_type != dmBuffer::VALUE_TYPE_FLOAT32 || stream_type_count != 3) {
+            dmLogError("Unsupported stream type for calculating bounds for culling. stream: '%s'", dmHashReverseSafe64(stream_name));
+            return;
+        }
```
Is this ok ?

https://github.com/defold/defold/issues/7110

otsakir committed:
Various fixes after PR review (#6445)

* Smarter algorithm to detect if buffer.boundsMin/Max are properly initialized. Introduces dmBuffer::IsAABBValid().
* Use default name "position" for vertex streams. Initialize in res_mesh.cpp
* Use Aos::maxPerElem() to calculate bounds with elegance :-)

otsakir committed:
Another round of changes (#6455)

* dmBuffer::Get/SetBounds() made public
* m_BoundingVersion field added to Buffer. It follows Buffer.m_ContentVersion when SetBounds() is used
* Added dmBuffer::BoundsUpToDate() to return if bounds are updated for a buffer (private)
* comp_mesh.cpp::UpdateMeshBounds() will calculate bounds and update buffer only if buffer bounds are not up to date.

JCash commented on engine/dlib/src/dlib/buffer.cpp:
```diff
@@ -397,6 +403,11 @@ namespace dmBuffer
         buffer->m_Stride = struct_size;
         buffer->m_ContentVersion = 0;
+        // meta data
+        buffer->boundsMin = dmVMath::Point3(1.0f, 1.0f, 1.0f);  // initialize to an "invalid" bounding box
+        buffer->boundsMax = dmVMath::Point3(-1.0f, -1.0f, -1.0f);
+        buffer->m_BoundingVersion = 999; // any value different than m_ContentVersion will allow triggering first bounds calculation
```
Perhaps set it to `contentversion + 1` to make it clear.

JCash replied:
👍

otsakir replied:
Yeah, it was my first thought too. 

But then i thought that when the buffer is updated and m_ContentVersion gets incremented (becomes 1) we'll have a case where `m_BoundingVersion  == m_ContentVersion` which is semantically wrong. Also, '-1' won't work since it's unisgned. 

I ended up using 999.

otsakir replied:
Another idea was to initialize set m_ContentVersion to 1 and m_BoundingVersion to 0. Semantically it seems the best. But i worried about breaking something.

otsakir commented:
The implementation has changed direction. I'm closing this PR to avoid confusion.
